DSL2 conversion

maxibor opened the pull request:
# nf-core/coproid pull request

This PR is the DSL2 conversion of nf-core/coproid, for releasing the versin 2.0 of the pipeline.
Most file changes are coming from nf-core boilerplate code (nf-core modules and configs), and probably need less attention to review. Workflow, documentations, and channels have more should probably looked closer at for reviewing.

## PR checklist

- [x] This comment contains a description of changes (with reason)
- [x] If you've fixed a bug or added code that should be tested, add tests!
- [x] Ensure the test suite passes (`nextflow run . -profile test,docker`).
- [x] Make sure your code lints (`nf-core lint .`).
- [x] Documentation in `docs` is updated
- [x] `CHANGELOG.md` is updated
- [x] `README.md` is updated

**Learn more about contributing:** [CONTRIBUTING.md](https://github.com/nf-core/coproid/tree/master/.github/CONTRIBUTING.md)

nf-core-bot committed:
Template update for nf-core/tools version 1.10.1

nf-core-bot committed:
Template update for nf-core/tools version 1.10.2

nf-core-bot committed:
Template update for nf-core/tools version 1.11

nf-core-bot committed:
Template update for nf-core/tools version 1.12

nf-core-bot committed:
Template update for nf-core/tools version 1.12.1

nf-core-bot committed:
Template update for nf-core/tools version 1.13

nf-core-bot committed:
Template update for nf-core/tools version 1.13.1

nf-core-bot committed:
Template update for nf-core/tools version 1.13.2

nf-core-bot committed:
Template update for nf-core/tools version 1.13.3

nf-core-bot committed:
Template update for nf-core/tools version 1.14

KevinMenden committed:
Template update for nf-core/tools version 2.0

nf-core-bot committed:
Template update for nf-core/tools version 2.0.1

nf-core-bot committed:
Template update for nf-core/tools version 2.1

nf-core-bot committed:
Template update for nf-core/tools version 2.2

nf-core-bot committed:
Template update for nf-core/tools version 2.3

nf-core-bot committed:
Template update for nf-core/tools version 2.3.1

nf-core-bot committed:
Template update for nf-core/tools version 2.3.2

nf-core-bot committed:
Template update for nf-core/tools version 2.4

nf-core-bot committed:
Template update for nf-core/tools version 2.5

nf-core-bot committed:
Template update for nf-core/tools version 2.5.1

nf-core-bot committed:
Template update for nf-core/tools version 2.6

nf-core-bot committed:
Template update for nf-core/tools version 2.7.1

nf-core-bot committed:
Template update for nf-core/tools version 2.7.2

nf-core-bot committed:
Template update for nf-core/tools version 2.8

nf-core-bot committed:
Template update for nf-core/tools version 2.9

nf-core-bot committed:
Template update for nf-core/tools version 2.10

nf-core-bot committed:
Template update for nf-core/tools version 2.11

nf-core-bot committed:
Template update for nf-core/tools version 2.11.1

nf-core-bot committed:
Template update for nf-core/tools version 2.12

nf-core-bot committed:
Template update for nf-core/tools version 2.13

nf-core-bot committed:
Template update for nf-core/tools version 2.13.1

nf-core-bot committed:
Template update for nf-core/tools version 2.14.0

nf-core-bot committed:
Template update for nf-core/tools version 2.14.1

maxibor committed:
restart of a blank state

maxibor committed:
nf-core base template

maxibor committed:
feat: add fastp module

MeriamOs committed:
add modules

MeriamOs committed:
add modules

MeriamOs committed:
changes coproid.nf and added subworkflows

MeriamOs committed:
added genome schema and changed nf-validation to nf-schema

MeriamOs committed:
pipeline is running, but only fastgc, fastp and multiqc

MeriamOs committed:
prepare genomes working, stuck at align_index variables

MeriamOs committed:
pipeline running, but modified main.nf from bowtie2 and samtools sort

MeriamOs committed:
fixed fasta input for samtools, but added .sorted to output bam - otherwise they have the same name

MeriamOs committed:
fixed double kraken2

MeriamOs committed:
added damageprofiler and start on quarto

MeriamOs committed:
added pydamage and some multiQC

MeriamOs committed:
added pydamage and some multiQC

maxibor committed:
add sam2lca build db subworkflow

MeriamOs committed:
fixed sourcepredict module embedding, added merge sam2lca

MeriamOs committed:
added pydamage merge and quarto workflow

MeriamOs committed:
quarto is now running, but need to change qmd file to improve report

MeriamOs committed:
Merge pull request #66 from nf-core/sam2lca_db_build

add sam2lca build db subworkflow

MeriamOs committed:
tidy up and moved sourcepredict into kraken subworkflow

MeriamOs committed:
patched modules

MeriamOs committed:
Merge branch 'dsl2' into quarto

MeriamOs committed:
Merge pull request #67 from MeriamOs/quarto

Merge quarto with dsl2

MeriamOs committed:
added damage profiler plots to report

MeriamOs committed:
make report prettier

MeriamOs committed:
changed test data, config and tidied up some code

MeriamOs committed:
added untar option for kraken database

MeriamOs committed:
added xz decompress for taxa_splite file, included normalisation and table in quartot report, and sorted test data

MeriamOs committed:
fixed sam2lca db channel, and taxa.sqlite.xz, tidying up

MeriamOs committed:
reporting, summary probability, fixed pydamage plot

MeriamOs committed:
updated samtools sort

MeriamOs committed:
trialing linting

MeriamOs committed:
fixing linting issues, updated modules and worksflows

MeriamOs committed:
new nf-core template, fixing linting issues

MeriamOs committed:
fixing linting, pre-commit issues

MeriamOs committed:
ignore logos

MeriamOs committed:
fixed some updated module issues

MeriamOs committed:
permission deniead issue of container on github

MeriamOs committed:
updatet nextflow version

MeriamOs committed:
reversed nextflow version

MeriamOs committed:
upgraded nextflow version

MeriamOs committed:
downgraded nf-core/schema

MeriamOs committed:
downgraded nf-core/schema

MeriamOs committed:
changed badge and ci.yml to match 23.10.0 nextflow version

MeriamOs committed:
trying to fix offending new-line value in bowtie2_build

MeriamOs committed:
trying to fix offending new-line value in bowtie2_build

MeriamOs committed:
changed process.shell in nextflow.config

MeriamOs committed:
test profile singularity

MeriamOs committed:
singularity profile set up

MeriamOs committed:
singularity profile set up

MeriamOs committed:
problem with quarto container on github test

MeriamOs committed:
problem with quarto container on github test

MeriamOs committed:
problem with quarto container on github test

MeriamOs committed:
problem with quarto container on github test

MeriamOs committed:
problem with quarto container on github test

MeriamOs committed:
problem with quarto container on github test

MeriamOs committed:
problem with quarto container on github test

MeriamOs committed:
added usage, output, citations

MeriamOs committed:
linting and pre-commit

MeriamOs committed:
added 'docker' to profiles

MeriamOs committed:
removed apptainer setup

MeriamOs committed:
docker container for quarto

MeriamOs committed:
docker container for quarto

MeriamOs committed:
output.md file

MeriamOs committed:
linting and pre-commit

maxibor committed:
Merge pull request #69 from nf-core/dsl2

Merge DSL2 with dev

MeriamOs committed:
started paper

MeriamOs committed:
git restore

MeriamOs committed:
file name issue

MeriamOs committed:
Template update for nf-core/tools version 3.2.0

MeriamOs committed:
Template update for nf-core/tools version 3.2.0

MeriamOs committed:
Template update for nf-core/tools version 3.2.0

MeriamOs committed:
Template update for nf-core/tools version 3.2.0

MeriamOs committed:
update template

MeriamOs committed:
Template update for nf-core/tools version 3.2.0

MeriamOs committed:
Template update for nf-core/tools version 3.2.0

MeriamOs committed:
Template update for nf-core/tools version 3.2.0

MeriamOs committed:
Template update for nf-core/tools version 3.2.0

MeriamOs committed:
Template update for nf-core/tools version 3.2.0

MeriamOs committed:
Template update for nf-core/tools version 3.2.0

MeriamOs committed:
Template update for nf-core/tools version 3.2.0

MeriamOs committed:
Template update for nf-core/tools version 3.2.0

MeriamOs committed:
Template update for nf-core/tools version 3.2.0

MeriamOs committed:
Template update for nf-core/tools version 3.2.0

MeriamOs committed:
Template update for nf-core/tools version 3.2.0

MeriamOs committed:
Template update for nf-core/tools version 3.2.0

MeriamOs committed:
Template update for nf-core/tools version 3.2.0

MeriamOs committed:
Template update for nf-core/tools version 3.2.0

MeriamOs committed:
full test

MeriamOs committed:
Template update for nf-core/tools version 3.2.0

MeriamOs committed:
Template update for nf-core/tools version 3.2.0

MeriamOs committed:
template update

MeriamOs committed:
fixed linting

MeriamOs committed:
fixed merge problem in main and full test data

MeriamOs committed:
fixed zenodo number and removed TODO statement

MeriamOs committed:
Restore main.nf to try and fix Module compilation error

MeriamOs committed:
remove latest everything testing

maxibor committed:
Merge pull request #71 from MeriamOs/dev

Template update

maxibor committed:
harmonize nf version requirement

maxibor committed:
remove redundant nf version constraint

maxibor committed:
add missing nf-core boilerplate param

maxibor committed:
update conda defintion

maxibor committed:
output directory locations cleanup

maxibor committed:
typo fixes

maxibor committed:
linting errors fix

maxibor committed:
nf ver >= 23.10.0

maxibor committed:
resourceLimits appears in 24.04.0

maxibor committed:
template should stay untouched

maxibor committed:
lint fix

maxibor committed:
update changelog

maxibor committed:
Merge pull request #72 from nf-core/hackathon_fixes

Hackathon fixes

maxibor committed:
cleanup: remove bbduk

maxibor committed:
changes suggested by @jen-reeve

maxibor committed:
reorder sections in multiqc

maxibor committed:
prettier

maxibor committed:
Merge pull request #75 from nf-core/reorder_multiqc

reorder sections in multiqc

jreeve-nv committed:
minor description update in schema

jreeve-nv committed:
updating changelog and contributers list

jreeve-nv committed:
removing TODO comments

jreeve-nv committed:
removing deprecated option from schema

MeriamOs committed:
Update quarto container

MeriamOs committed:
changed location of samplesheet and genomesheet to test-datasets

MeriamOs committed:
changed location of samplesheet and genomesheet to test-datasets

MeriamOs committed:
Minor quarto report changes

This fixes "squished" damageplots if more than a couple of samples are processed

jen-reeve committed:
Update subworkflows/local/utils_nfcore_coproid_pipeline/main.nf

Co-authored-by: Nicolas Vannieuwkerke <101190534+nvnieuwk@users.noreply.github.com>

maxibor committed:
Merge pull request #77 from jen-reeve/json-schema-improvements

JSON schema improvements

maxibor committed:
Merge pull request #78 from nf-core/MeriamOs-patch-1

Fix quarto container and location of input files

maxibor committed:
update minimum nf version

maxibor committed:
bump coproid version to stable 2.0

maxibor committed:
update version

maxibor committed:
fix contributor map

nvnieuwk commented on CHANGELOG.md:
```diff
@@ -1,44 +1,31 @@
+and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).
 
-### Fixed
+## v2.0dev - [date]
```
Can you remove the `dev` here?

nvnieuwk replied:
versions should also follow the semantic versioning, so `2.0.0` is the correct version here

nvnieuwk commented on assets/_extensions/nf-core/coproid_logo.png:
can you make the last two letters of the name lowercase?

nvnieuwk commented on assets/schema_genomes.json:
```diff
@@ -0,0 +1,67 @@
+            "genome_name": {
+                "type": "string",
+                "pattern": "^\\S+$",
+                "errorMessage": "Genome name must be provided and cannot contain spaces",
```
```suggestion
                "errorMessage": "Genome name must be provided as a string value and cannot contain spaces",
```

nvnieuwk commented on assets/schema_genomes.json:
```diff
@@ -0,0 +1,67 @@
+            "taxid": {
+                "type": "integer",
+                "pattern": "^\\d+$",
+                "errorMessage": "Taxid must be provided",
```
```suggestion
                "errorMessage": "Taxid must be provided as an integer value",
```

nvnieuwk commented on assets/schema_genomes.json:
```diff
@@ -0,0 +1,67 @@
+            "genome_size": {
+                "type": "integer",
+                "pattern": "^\\d+$",
+                "errorMessage": "Genome size must be provided and cannot contain spaces",
```
```suggestion
                "errorMessage": "Genome size must be provided as an integer value and cannot contain spaces",
```

nvnieuwk commented on assets/schema_genomes.json:
```diff
@@ -0,0 +1,67 @@
+                "meta": ["genome_size"]
+            },
+            "igenome": {
+                "anyOf": [
+                    {
+                        "type": "string",
+                        "format": "file-path",
+                        "pattern": "^\\S+$"
+                    },
+                    {
+                        "type": "string",
+                        "maxLength": 0
+                    }
+                ],
+                "exists": true,
+                "errorMessage": "iGenome name or fasta must be provided"
```
```suggestion
                "type": "string",
                "format": "file-path",
                "pattern": "^\\S+$",
                "exists": true,
                "errorMessage": "iGenome name or fasta must be an existing file path that does not contain spaces"
```

nvnieuwk replied:
This was unnecessarily complicated, This does exactly the same thing and looks a lot cleaner :)

nvnieuwk commented on assets/schema_genomes.json:
```diff
@@ -0,0 +1,67 @@
+                "errorMessage": "iGenome name or fasta must be provided"
+            },
+            "fasta": {
+                "anyOf": [
+                    {
+                        "type": "string",
+                        "format": "file-path",
+                        "pattern": "^\\S+\\.(fasta|fas|fa|fna)(\\.gz)?$"
+                    },
+                    {
+                        "type": "string",
+                        "maxLength": 0
+                    }
+                ],
+                "exists": true,
+                "errorMessage": "iGenome name or fasta must be provided"
```
```suggestion
                "type": "string",
                "format": "file-path",
                "pattern": "^\\S+\\.(fasta|fas|fa|fna)(\\.gz)?$",
                "exists": true,
                "errorMessage": "Fasta must be an existing file that ends with '.fasta', '.fai', '.fna' or '.fa'. This file can be gzipped"
```

nvnieuwk commented on assets/schema_input.json:
```diff
@@ -0,0 +1,33 @@
+            "sample": {
+                "type": "string",
+                "pattern": "^\\S+$",
+                "errorMessage": "Sample name must be provided and cannot contain spaces",
```
```suggestion
                "errorMessage": "Sample name must be provided as a string value and cannot contain spaces",
```

nvnieuwk commented on assets/schema_input.json:
```diff
@@ -0,0 +1,33 @@
+                "format": "file-path",
+                "exists": true,
+                "pattern": "^\\S+\\.f(ast)?q\\.gz$",
+                "errorMessage": "FastQ file for reads 1 must be provided, cannot contain spaces and must have extension '.fq.gz' or '.fastq.gz'"
```
```suggestion
                "errorMessage": "FastQ file for reads 1 must be provided, cannot contain spaces, has to exist and must have extension '.fq.gz' or '.fastq.gz'"
```

nvnieuwk commented on assets/schema_input.json:
```diff
@@ -0,0 +1,33 @@
+                "format": "file-path",
+                "exists": true,
+                "pattern": "^\\S+\\.f(ast)?q\\.gz$",
+                "errorMessage": "FastQ file for reads 2 cannot contain spaces and must have extension '.fq.gz' or '.fastq.gz'"
```
```suggestion
                "errorMessage": "FastQ file for reads 2 cannot contain spaces, has to exist and must have extension '.fq.gz' or '.fastq.gz'"
```

nvnieuwk commented on conf/modules.config:
```diff
@@ -0,0 +1,135 @@
+    withName: QUARTONOTEBOOK {
+        ext.prefix = "${params.file_prefix}_report"
+        singularity.registry  = 'cloud.sylabs.io'
+        apptainer.registry    = 'cloud.sylabs.io'
```
What's the point of having these registries here? The module seems to use Seqera containers :)

nvnieuwk commented on modules.patch:
Can you please push most of these changes to the official nf-core modules. There's some very good stuff in here that everyone could benefit from

MeriamOs replied:
Haven't done this yet, but can do tomorrow (NZ time)!

nvnieuwk commented on nextflow.config:
```diff
@@ -1,170 +1,308 @@
+    kraken2_db                 = null
+    sp_sources                 = null
+    sp_labels                  = null
+    taxa_sqlite                = "params.modules_testdata_base_path + /genomics/prokaryotes/metagenome/taxonomy/misc/taxa_sqlite.xz"
+    taxa_sqlite_traverse_pkl   = "params.modules_testdata_base_path + /genomics/prokaryotes/metagenome/taxonomy/misc/taxa_sqlite_traverse.pkl"
```
This doesn't seem right, I think you missplaced a quote. The `modules_testdata_base_path` won't be used here

nvnieuwk commented on nextflow.config:
```diff
@@ -1,170 +1,308 @@
+    mainScript      = 'main.nf'
+    defaultBranch   = 'master'
+    nextflowVersion = '!>=24.10.0'
+    version         = '2.0'
```
```suggestion
    version         = '2.0.0'
```

nvnieuwk commented on subworkflows/local/kraken2_classification/main.nf:
```diff
@@ -0,0 +1,103 @@
+include { UNTAR           } from '../../../modules/nf-core/untar/main'
+include { XZ_DECOMPRESS   } from '../../../modules/nf-core/xz/decompress/main'
+
+if (params.sp_sources  )              { ch_sp_sources = file(params.sp_sources) } else { error("SourcePredict sources file not specified!") }
+if (params.sp_labels   )              { ch_sp_labels  = file(params.sp_labels) } else { error("SourcePredict labels file not specified!") }
+if (params.taxa_sqlite_traverse_pkl ) { ch_sqlite_traverse = file(params.taxa_sqlite_traverse_pkl) } else { error("Ete3 taxa.sqlite.traverse file not specified!") }
```
Can you add the following checks to these parameters in `nextflow_schema.json`?
```
"format": "file-path"
"exists": true
```

You can remove these checks completely then as nf-schema will fully cover this

nvnieuwk commented on workflows/coproid.nf:
```diff
@@ -0,0 +1,292 @@
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+*/
+
+if (params.kraken2_db ) { ch_kraken2_db = file(params.kraken2_db) } else { error("Kraken2 database path not specified!") }
```
This check if also already covered by nf-schema :)

nvnieuwk commented on workflows/coproid.nf:
```diff
@@ -0,0 +1,292 @@
+        ALIGN_INDEX.out.bai
+    ).map {
+        meta, bam, bai -> [['id':meta.sample_name], bam] // meta.id, bam
+    }.groupTuple()
```
This will create a bottleneck in your pipeline. The pipeline will wait until all FASTQ files are aligned before actually performing the `.groupTuple` (which could cause significant slowdowns). 

You can fix this by telling the operator how big the expected groups should be. You can do this in two ways:
1. All groups will be the same size? Use the `size` option for the [operator](https://www.nextflow.io/docs/latest/reference/operator.html)
2. Group sizes are dynamic? Use the `groupKey` function to dynamically assign a group size per group key (This can also be found in the [.groupTuple](https://www.nextflow.io/docs/latest/reference/operator.html) docs

MeriamOs replied:
Thanks for this @nvnieuwk! I think I have fixed that now with the size option. 
I'm not sure if I've done it the right way, but I made all the changes here https://github.com/nf-core/coproid/pull/79
I don't have direct writing access to dev.

nvnieuwk replied:
I approved the PR :) I'm not familiar with this pipeline so I don't know what the best solution is here. 

In your implementation, you told the pipeline that all channel entries after the `.groupTuple` should consist of the merge of two channel entries with the same ID.

maxibor replied:
I don't know if the bottleneck is a big issue, as we anyway generate a single report per run, where the results of all samples are needed.

MeriamOs replied:
@maxibor true. I guess the size operator will at least stop the bottlenecking before the quarto reporting? 

I changed it to .groupTuple(size: 2) - each tuple will contain the bam and the bai file, so I think this is right. I tested it with the test data, and it ran smoothly. I that looks good to you, I think we can merge https://github.com/nf-core/coproid/pull/79 ?

nvnieuwk reviewed (changes requested):
A lot of minor comments and one serious bottleneck that needs to be solved, but looking very good! Great job!

MeriamOs committed:
Update version + removed spelling mistake

MeriamOs committed:
Fixed review changed in genome schema

MeriamOs committed:
Fixed review changes in input schema

MeriamOs committed:
Removed some left over code from previous container issues

MeriamOs committed:
Fixed quotes and version

MeriamOs committed:
Update nextflow_schema.json for kraken2 files

MeriamOs committed:
Removed tests already covered in nextflow schema

MeriamOs committed:
Removed tests already covered in nextflow schema

MeriamOs committed:
Removed tests already covered in nextflow schema

MeriamOs committed:
Removed tests already covered in nextflow schema

MeriamOs committed:
Removed bbduk patch as this module has been removed

MeriamOs committed:
Updated logo

MeriamOs committed:
Updated logo

MeriamOs committed:
Delete assets/_extensions/nf-core/coproid_logo.png

MeriamOs committed:
Updated logo

MeriamOs committed:
Updated logo

MeriamOs committed:
Delete assets/_extensions/nf-core/coproid_logo1.png

MeriamOs committed:
Fixed bottleneck for bam files

MeriamOs committed:
Fixed quotes

MeriamOs committed:
Removed tests already covered in nextflow schema

MeriamOs committed:
Fix taxa.sqlite files

MeriamOs committed:
version harmonisation

MeriamOs committed:
version harmonisation

MeriamOs committed:
pre-commit

itrujnara commented on modules/local/create_acc2tax/main.nf:
```diff
@@ -0,0 +1,22 @@
+    tuple val(meta), path(fasta)
+
+    output:
+    path("*.accession2taxid"), emit: acc2tax
```
Does this follow some common format? If so, an extension would be nice

MeriamOs replied:
The extension is .accession2taxid

itrujnara replied:
Pretty sure it could be expressed as a more common format (e.g. CSV), but I'll just leave it for your later consideration

itrujnara commented on modules/local/create_acc2tax/main.nf:
```diff
@@ -0,0 +1,22 @@
+process CREATE_ACC2TAX {
+    tag "${meta.genome_name}"
+    label 'process_single'
```
You may consider creating an extra label for this process and a few similar ones, as `process_single` still requests substantial resources (especially time)

MeriamOs replied:
Thanks for this suggestion, I have created a process_short like you have for reportho :)

itrujnara commented on modules/local/create_acc2tax/main.nf:
```diff
@@ -0,0 +1,22 @@
+
+    """
+    create_acc2tax.py $fasta -t ${meta.taxid}
+    """
```
Please add stubs to modules to enable stub tests

MeriamOs replied:
Do local modules need stub tests? I don't see them in a lot of other pipelines

itrujnara commented on modules/local/sam2lca/prep_db/main.nf:
```diff
@@ -0,0 +1,25 @@
+    path(acc2tax)
+
+    output:
+    path("*.md5"), emit: acc2tax_md5
```
Harshil alignment

itrujnara commented on modules/local/sam2lca/updatedb/main.nf:
```diff
@@ -0,0 +1,62 @@
+        path(acc2tax_md5) // acc2tax.gz.md5
+
+    output:
+        path "sam2lca_db" , emit: sam2lca_db
```
Harshil alignment

itrujnara commented on modules/local/sam2lca/updatedb/main.nf:
```diff
@@ -0,0 +1,62 @@
+        task.ext.when == null || task.ext.when
+
+    script:
+        def args = task.ext.args ?: ''
```
Harshil alignment

itrujnara commented on modules/local/sam2lca/updatedb/meta.yml:
```diff
@@ -0,0 +1,80 @@
+name: "sam2lca_updatedb"
```
Looks like a module ready for the nf-core modules repo, are you making a PR there? If so, please finalize and replace this with the nf-core version

itrujnara commented on modules/local/sam2lca/updatedb/tests/main.nf.test:
```diff
@@ -0,0 +1,35 @@
+    tag "sam2lca"
+    tag "sam2lca/updatedb"
+
+    test("test-sam2lca-updatedb - test dataset") {
```
Going into proper module review mode
There should be at least one stub test, as well as tests for other input conmbinations

itrujnara commented on modules/local/sam2lca/updatedb/tests/main.nf.test:
```diff
@@ -0,0 +1,35 @@
+
+        then {
+            assertAll(
+                { assert process.success }
```
There should be a maximally strict assertion on the output, ideally `assert snapshot(process.out).match()`

itrujnara commented on subworkflows/local/align_index/main.nf:
```diff
@@ -0,0 +1,27 @@
+        ch_versions = ch_versions.mix(SAMTOOLS_INDEX.out.versions.first())
+
+    emit:
+        bam =  BOWTIE2_ALIGN.out.bam
```
Harshil alignment

itrujnara commented on subworkflows/local/kraken2_classification/main.nf:
```diff
@@ -0,0 +1,103 @@
+            false
+        )
+        ch_versions = ch_versions.mix(KRAKEN2_KRAKEN2.out.versions.first())
+        ch_versions = ch_versions.mix(UNTAR.out.versions.first())
```
Move this into the if block, the process is not guaranteed to run

itrujnara commented on subworkflows/local/kraken2_classification/main.nf:
```diff
@@ -0,0 +1,103 @@
+
+        KRAKEN_MERGE(kraken_read_count)
+
+//    ch_versions = ch_versions.mix(KRAKEN_MERGE.out.versions.first())
```
Why are these versions not emitted?

itrujnara commented on subworkflows/local/kraken2_classification/main.nf:
```diff
@@ -0,0 +1,103 @@
+            sqlite = file(params.taxa_sqlite)
+    }
+
+    ch_versions = ch_versions.mix(XZ_DECOMPRESS.out.versions.first())
```
This is not guaranteed to run

itrujnara commented on subworkflows/local/kraken2_classification/main.nf:
```diff
@@ -0,0 +1,103 @@
+    ch_versions = ch_versions.mix(SOURCEPREDICT.out.versions.first())
+
+    emit:
+        sp_report = SOURCEPREDICT.out.report
```
Harshil alignment

itrujnara commented on subworkflows/local/merge_sort_index_samtools/main.nf:
```diff
@@ -0,0 +1,31 @@
+    ch_versions = ch_versions.mix(SAMTOOLS_INDEX.out.versions.first())
+
+    emit:
+    bam = SAMTOOLS_SORT.out.bam
```
Harshil alignment

itrujnara commented on subworkflows/local/prepare_genome_indices/main.nf:
```diff
@@ -0,0 +1,58 @@
+    ch_versions = ch_versions.mix(BOWTIE2_BUILD.out.versions.first())
+
+    emit:
+    genomes = ch_genomes
```
Harshil alignment

itrujnara commented on subworkflows/local/quarto_reporting/main.nf:
```diff
@@ -0,0 +1,48 @@
+    ch_versions = ch_versions.mix(QUARTONOTEBOOK.out.versions.first())
+
+    emit:
+    versions          = ch_versions
```
Nothing to align to, this equal sign can be closer

itrujnara commented on subworkflows/local/sam2lca_db/main.nf:
```diff
@@ -0,0 +1,44 @@
+            SAM2LCA_PREPDB.out.acc2tax_md5
+        )
+
+//        ch_versions = ch_versions.mix(CREATE_ACC2TAX.out.versions.first())
```
Why no versions?

itrujnara commented on workflows/coproid.nf:
```diff
@@ -0,0 +1,292 @@
+
+    DAMAGEPROFILER_MERGE(
+        damageprofiler_reports
+    )
```
Missing versions

itrujnara commented on workflows/coproid.nf:
```diff
@@ -0,0 +1,292 @@
+
+    PYDAMAGE_MERGE (
+        pydamage_reports
+    )
```
Missing versions here as well

itrujnara commented on workflows/coproid.nf:
```diff
@@ -0,0 +1,292 @@
+
+    main:
+
+    ch_versions = Channel.empty()
```
Harshil alignment everywhere

itrujnara commented on workflows/coproid.nf:
```diff
@@ -0,0 +1,292 @@
+                [],
+                [],
+                []
+            )
```
Missing versions

itrujnara commented on workflows/coproid.nf:
```diff
@@ -0,0 +1,292 @@
+
+    SAM2LCA_MERGE (
+        sam2lca_reports
+    )
```
Missing versions

itrujnara commented on workflows/coproid.nf:
```diff
@@ -0,0 +1,292 @@
+    //
+    QUARTO_REPORTING (
+        ch_quarto
+    )
```
Missing versions

itrujnara commented on workflows/coproid.nf:
```diff
@@ -0,0 +1,292 @@
+        []
+    )
+
+    emit:multiqc_report = MULTIQC.out.report.toList() // channel: /path/to/multiqc_report.html
```
Missing newline

itrujnara reviewed (commented):
Rewriting DSL1 pipelines is always a great contribution. You have done a very solid job with this pipeline.
I have found several technicalities to fix or to discuss along the way. I also have 2 general remarks:
– For CSV file handling, `csvtk` modules are a great tool, and you should consider them; for complex tasks, you should consider using `polars`, as it is orders of magnitude faster than `pandas` in many cases
– It is useful to mix versions immediately after the process invocation, as it makes it easier to track whether all versions are correctly handled; and every module should emit versions if it uses any dependency, even Python or coreutils

MeriamOs committed:
Fix version from dev to 2.0.0

MeriamOs committed:
Add coproid workflow figure

MeriamOs committed:
Added coproid workflow figure to README.md

MeriamOs commented on README.md:
```diff
@@ -1,115 +1,131 @@
 
-A detailed description of coproID can be found in the [article published in PeerJ](https://peerj.com/articles/9001).
+(A) First coproID performs comparative mapping of all reads agains two (or three) target genomes (genome1, genome2, and potentially genome3) and computes a host-DNA species ratio (NormalisedProportion).
+(B) Then coproID performs metagenomic taxonomic profiling, and compares the obtained profiles to modern reference samples of the target species metagenomes. Using machine learning, coproID then estimates the host source from the metagenomic taxonomic composition (SourcepredictProportion).
```
Added the workflow figure. Had created it a while ago, but forgot to add haha - thanks for picking that up!

MeriamOs committed:
Add MIT license

MeriamOs committed:
Add MIT license

MeriamOs committed:
Add MIT license

MeriamOs committed:
Add MIT license

MeriamOs committed:
Add MIT license

MeriamOs committed:
Add module docstring

MeriamOs committed:
Add module docstring

MeriamOs committed:
Add module docstring

MeriamOs committed:
Add module docstring

MeriamOs committed:
Add module docstring

MeriamOs committed:
Add module docstring

MeriamOs committed:
Add module docstring

MeriamOs committed:
Add module docstring

MeriamOs committed:
Add module docstring

MeriamOs committed:
Add module docstring

MeriamOs committed:
Add module docstring

MeriamOs commented on bin/damageprofiler_merge.py:
```diff
@@ -0,0 +1,83 @@
+        file_parts = file_path.split("/")
+
+        # Check if we have enough parts in the path
+        if len(file_parts) < 2:
```
This should be valid - damageprofiler creates output files within sample directories. All the files within each directory have the exact same name, this is the standard way of outputting by damageprofiler. That's why we use the directory's name

MeriamOs commented on bin/kraken_merge.py:
```diff
@@ -1,59 +1,63 @@
+    return basename
 
 
 def merge_csv(all_csv):
```
Will consider this for the next release

MeriamOs committed:
Harshil alignment

MeriamOs committed:
Fixed missing new line

MeriamOs committed:
added missing versions

MeriamOs commented on bin/pydamage_merge.py:
```diff
@@ -0,0 +1,43 @@
+import sys
+import pathlib
+
+def process_file(file_path):
```
I'll consider this for the next release

MeriamOs committed:
Hardcoded db_name

MeriamOs commented on bin/sam2lca_merge.py:
```diff
@@ -0,0 +1,45 @@
+
+import sys
+import os
+import pandas as pd
```
Will consider this for next release

MeriamOs commented on bin/sam2lca_json.py:
```diff
@@ -0,0 +1,36 @@
+    return parser.parse_args()
+
+
+def write_json(acc2taxid, md5, db_name="adnamap"):
```
This script was used for another pipeline where you can set the db_name. Hardcoded the value now

MeriamOs committed:
added short process label to save time requests

MeriamOs committed:
missing version

MeriamOs committed:
harshil alignment

MeriamOs committed:
missing version

MeriamOs committed:
harshil alignment

MeriamOs committed:
fix verions kraken2 classification

MeriamOs committed:
fix verions kraken2 classification

MeriamOs committed:
removed whitespaces

MeriamOs commented:
> Rewriting DSL1 pipelines is always a great contribution. You have done a very solid job with this pipeline. I have found several technicalities to fix or to discuss along the way. I also have 2 general remarks: – For CSV file handling, `csvtk` modules are a great tool, and you should consider them; for complex tasks, you should consider using `polars`, as it is orders of magnitude faster than `pandas` in many cases – It is useful to mix versions immediately after the process invocation, as it makes it easier to track whether all versions are correctly handled; and every module should emit versions if it uses any dependency, even Python or coreutils

Thanks for you great review @itrujnara ! I have made all the edits in https://github.com/nf-core/coproid/pull/80
- nf-core workflow figure has been added. 
- I have fixed all the harshil alignments and versions. 
- I added a process_short label for modules requiring few resources. 
- I will definitely consider those tools for CSV handling for the next release. 
- I have left the sam2lca/tests for a potential nf-core module  as Maxime has been working on this.

MeriamOs committed:
added stubs to local modules

MeriamOs committed:
fixed doc string to imperative mood

MeriamOs committed:
changed logo for quarto report

maxibor committed:
Merge pull request #80 from nf-core/MeriamOs-review2-fixes

Meriam van Os review2 fixes

nvnieuwk reviewed (approved):
LGTM!

itrujnara commented on CHANGELOG.md:
```diff
@@ -3,10 +3,10 @@
 and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).
 
-## v2.0dev - [date]
+## v2.0.0 - 03/04/2025
```
Update this before merging

itrujnara commented on README.md:
```diff
@@ -23,10 +23,13 @@
-(B) Then coproID performs metagenomic taxonomic profiling, and compares the obtained profiles to modern reference samples of the target species metagenomes. Using machine learning, coproID then estimates the host source from the metagenomic taxonomic composition (SourcepredictProportion).
+1. First coproID performs comparative mapping of all reads agains two (or three) target genomes (genome1, genome2, and potentially genome3) and computes a host-DNA species ratio (NormalisedProportion).
+2. Then coproID performs metagenomic taxonomic profiling, and compares the obtained profiles to modern reference samples of the target species metagenomes. Using machine learning, coproID then estimates the host source from the metagenomic taxonomic composition (SourcepredictProportion).
+3. Finally, coproID combines A and B proportions to predict the likely host of the metagenomic sample.
```
There is no A and B anymore?

itrujnara reviewed (approved):
Good job! Don't forget to update the release date before you press the button

MeriamOs committed:
Update date in CHANGELOG.md

MeriamOs committed:
Minor chnages README.md

Fix inconsistency in numbering, and added ancient DNA damage estimation to description

MeriamOs committed:
Update date in CHANGELOG.md

MeriamOs committed:
Merge pull request #82 from nf-core/MeriamOs-for-release

Last minor changes for release
